### PR TITLE
fix(select,deps,delegation): follow-up fixes stranded past #129's merge

### PR DESCRIPTION
**Pull request by senamakel**
## Summary

Follow-up to #129, which merged at head `674b782` while review fixes were still
landing on the same branch. 15 commits are stranded past that merge point and
are not on `main`. They are all fixes for findings raised by CodeRabbit, Codex
and tinysweeper against #129's own tip, each verified before push.

The most important one is a **ranking regression introduced during #129's own
review**, caught by the parity snapshot that PR added.

## API Or Behavior Changes

**`harness::tool::select` — ranking regression fixed (the headline).**

#129 review moved the ambiguous resource nouns (`email` / `message` / `dm`) out
of `ToolVerb::Send`'s alias list into a separate `SEND_NOUN_ALIASES`, gated on
`found.is_empty()`. The intent was sound — `"read email"` should be `Read`, not
`Read + Send` — but the gate is not equivalent to the behaviour it replaced:

```
"Post a message to the #general channel saying the deploy is complete"
  → "post" matches a Create alias  → found = {Create}, non-empty
  → the Send noun fallback never runs
  → SLACK_CREATE_CHANNEL, SLACK_CREATE_CHANNEL_BASED_CONVERSATION, SLACK_SCHEDULE_MESSAGE …
  → SLACK_SEND_MESSAGE not in the top 15 at all
```

Fixed by gating the noun fallback on *conflict* rather than on emptiness: the
noun still applies alongside `Create` (posting/writing/drafting a message is a
send intent expressed with a creation verb), and still does not apply alongside
`Read`/`List`/`Update`/`Delete`/`Merge`.

This keeps **both** properties, which is why it is preferable to reverting:

- `resource_noun_does_not_add_send_alongside_an_explicit_conflicting_verb`
  (added during #129 review) still passes — `"read email"` → `{Read}`,
  `"delete a message"` → `{Delete}`.
- The host's `adapter_ranking_matches_the_pre_extraction_snapshot` passes
  again — orderings captured from the pre-extraction code over a 1,000-action
  real catalogue, which is the parity contract the extraction was built on.

**`Cargo.toml` — `base64` pinned back to `0.22`.** Dependabot #125 moved this
crate to `0.23.1`, but the whole `reqwest` → `hyper-rustls` → `hyper-util` chain
still resolves `0.22.1`, so the crate now pulled a *second* copy. In OpenHuman
that tripped the kernel-dependency-floor ratchet (`288 → 289` packages, names
and native builds unchanged — the signature of a duplicate, not a new dep).
Verified back at `288 / 270 / 2` with a single `base64 v0.22.1` in the graph.
This partially reverts #125's intent for this crate; dependabot will re-propose
`0.23` once reqwest's tree moves, which is fine.

**`graph::delegation` — two real concurrency/schema holes closed.**
`resume_delegation` is a public entry point that bypassed
`run_or_resume_delegation`'s per-thread lock and schema check entirely: it now
serialises on the same lock and rejects a schema-mismatched checkpoint instead
of resuming against it. A cancellation check was also missing at the durable
approval interrupt node, so a cancelled run could route into approval rather
than finalize.

**`graph::dag` — documentation only.** `has_cycle` deliberately builds the graph
from the first declaration of a repeated id. The docs now state the
*consequence*, not just the rationale: a caller that wants to add edges to an
existing node must merge them into that node's single declaration, because
appending a second declaration silently drops those edges — and with them any
cycle they would have closed.

## Tests

- [x] `cargo fmt --check`
- [x] `cargo clippy --all-targets -- -D warnings`
- [x] `cargo clippy --all-targets --all-features -- -D warnings`
- [x] `cargo build --all-targets`
- [x] `cargo build --all-targets --all-features`
- [x] `cargo test` — **2,520 passed / 0 failed**
- [x] `cargo test --all-features` — **2,653 passed / 0 failed**

Cross-checked against the consumer, since the regression was only visible there:
OpenHuman's `tool_filter` suite is green on this tip (11 passed / 0 failed),
including `adapter_ranking_matches_the_pre_extraction_snapshot`,
`real_data_slack_send_message` and `repro_3152_create_page_reachable_in_top_k`.
Kernel floor re-measured at `288 / 270 / 2` against `scripts/kernel-floor.limits`.

New tests here: schema-mismatched-checkpoint rejection and resume serialisation
for `resume_delegation`, cancellation-at-approval-node routing, and the
schema-tag scoping check in `graph::checkpoint`.

## Documentation

`has_cycle`'s duplicate-id consequence (above) and an expanded rationale on
`detect_verbs` explaining precisely why `found.is_empty()` is not equivalent to
the pre-extraction behaviour, so the same regression is not reintroduced.

## Related

Follow-up to #129. Companion consumer PR: tinyhumansai/openhuman#5852, which
pins this branch by gitlink and cannot go green until this merges — its
`Rust Feature-Gate Smoke` lane fails on the duplicate `base64` and its
`Rust Core Coverage` lane fails on the ranking regression, both fixed here.

## Summary by CodeRabbit

* **Bug Fixes**
  * Cancellations during approval pauses now finalize correctly without triggering unintended retries or resumes.
  * Concurrent delegation resumes for the same thread are handled safely.
  * Incompatible or malformed checkpoints are rejected with clearer error classification.
  * Tool detection no longer incorrectly interprets resource nouns as a “send” action when another explicit action is present.

* **Documentation**
  * Clarified cycle-detection behavior when duplicate node identifiers are encountered.

### CODE REVIEW

File: src/graph/delegation/test.rs [hunk 1]
```diff
@@ -824,6 +824,178 @@ async fn cancelled_checkpoint_still_scheduling_finalize_is_resumed_not_terminal(
...
+    run_delegation_durable(make_config(), make_runner())
+        .await
+        .expect("parks on approval")
+        .pending
+        .expect("parked on the approval interrupt");
+
+    let (r1, r2) = tokio::join!(
+        resume_delegation(make_config(), json!("approve_once"), make_runner()),
+        resume_delegation(make_config(), json!("approve_once"), make_runner()),
+    );
```
[the remark this answers is not in the document]
Comment (senamakel) [reply src/graph/delegation/test.rs:976]: Confirmed and fixed. The critique was exactly right: the approval interrupt's resume path routes straight from `approval` to `finalize` without ever calling `run_stage` (`finalize` only reads accumulated state), so the old `run_stage`-based counter only ever observed the SETUP `run_delegation_durable` call -- the assertion passed even with the per-thread lock removed entirely.

Rewrote the test to wrap the checkpointer with a `BlockingGetCheckpointer` that instruments `get()` (which IS on `resume_delegation`'s critical path -- the schema-mismatch check reads the checkpoint before dispatching) with a concurrency counter plus artificial delay, delegating every other `Checkpointer` method to the real `FileCheckpointer`.

Verified this actually tests something: manually removed the `thread_lock`/`_guard` acquisition in `resume_delegation` and reran -- the test now fails correctly (assertion `left == right` failed, left: 2, right: 1), confirming it would have caught the exact regression this test exists to prevent. Restored the lock and the suite is green again (31/31 delegation tests).

File: src/harness/tool/select/mod.rs [hunk 2]
```diff
@@ -173,6 +186,39 @@ fn detect_verbs(prompt: &str) -> HashSet<ToolVerb> {
...
+    // email", "delete a message" are Read/Delete, not Send). It does apply
+    // alongside `Create`, because "post/write/draft a message" is a send
+    // intent expressed with a creation verb — and suppressing it there is
+    // what dropped SLACK_SEND_MESSAGE out of the top 15 for "Post a message
+    // to #general". Pinned by the host's pre-extraction ranking snapshot.
+    let conflicts_with_send = found.iter().any(|v| {
+        matches!(
+            v,
+            ToolVerb::Read | ToolVerb::List | ToolVerb::Update | ToolVerb::Delete | ToolVerb::Merge
+        )
```
[the remark this answers is not in the document]
Comment (senamakel) [reply src/harness/tool/select/mod.rs:212]: Confirmed and fixed at 8f804a8. The finding was correct, and I want to be explicit about how I verified it given the last round's regression came from exactly this kind of change.

Root cause: the previous conflict check suppressed the Send-resource-noun fallback whenever ANY conflicting verb was found ANYWHERE in the prompt, regardless of distance. For "find Alice and email her the report" that suppressed Send because "find" (a List alias) is present, even though "find Alice" and "email her the report" are two independent instructions joined by "and" -- and the verb gate FILTERS OUT (not just down-ranks) a tool whose verb isn't in the query's detected verb set, so GMAIL_SEND_EMAIL would have been dropped entirely rather than merely deprioritized.

Fix: replaced the whole-prompt conflict check with a same-clause check. A conflicting verb only suppresses a Send-resource-noun when no clause-boundary word ("and", "then", "also", "plus", "next") separates them in the token stream -- i.e. when the noun reads as that verb's direct object ("read email", "delete that old email") rather than a second, independent instruction. Distance alone cannot distinguish these two shapes: "delete THAT OLD email" and "find Alice AND email her" both have exactly two words between the verb and the noun; only the presence of "and" signals a second clause. My first attempt used a fixed word-distance window and I caught it failing exactly this way before pushing -- it correctly handled the compound case but then failed "delete that old email" (2 modifier words, same distance as the compound example), which is why the final fix keys on the boundary word instead of distance.

Regression tests added: `compound_task_preserves_send_alongside_a_distant_unrelated_verb` (the exact example from the finding, plus "delete that old email" as the boundary case it must still suppress).

Cross-checked against the consumer per the standing instruction for this area: OpenHuman's `tool_filter` suite is green on this tip (11 passed / 0 failed), including `adapter_ranking_matches_the_pre_extraction_snapshot`, `real_data_slack_send_message`, `real_data_gmail_send_email`, and `real_data_gmail_delete_emails` -- so this does not reintroduce the ranking regression this whole area exists to prevent, and does not regress the Read/Delete-conflict cases either.

File: src/harness/tool/select/test.rs [hunk 3]
```diff
@@ -154,6 +154,21 @@ fn verb_detection_handles_aliases() {
...
+    // resource noun — only the explicit action verb should be present.
+    let v = detect_verbs("read email");
+    assert_eq!(v, HashSet::from([ToolVerb::Read]));
+
+    let v = detect_verbs("delete a message");
+    assert_eq!(v, HashSet::from([ToolVerb::Delete]));
+
+    // No explicit verb at all: the resource noun alone may still imply Send.
+    let v = detect_verbs("message support channel");
+    assert!(v.contains(&ToolVerb::Send));
```
[the remark this answers is not in the document]
Comment (senamakel) [reply src/harness/tool/select/test.rs:175]: Fixed. Replaced the permissive assertion with an exact HashSet comparison: `assert_eq!(v, HashSet::from([ToolVerb::Create, ToolVerb::Send]))` for "post a message to general channel", so an implementation that suppresses Send whenever any verb is found can no longer pass this test silently.
